### PR TITLE
Space Travelers' Hub | React & Redux Group Project

### PR DESCRIPTION
**Pull request by ghost**
## Project description
In this project, @lazyCode-404, @zakiyousufi and @username-forGithub worked with data from the SpaceX API ([Rockets](https://api.spacexdata.com/v3/rockets) & [Missions](https://api.spacexdata.com/v3/missions) & [Dragons](https://api.spacexdata.com/v3/dragons)). We built a web application for a company that provides commercial and scientific space travel services. The application allows users to book rockets and join selected space missions.

### Description
- Setup App using React and Redux
- Install dependencies including [React Redux](https://react-redux.js.org/). [Redux Logger](https://www.npmjs.com/package/redux-logger), and [React Router](https://reactrouter.com/web/guides/quick-start)
- Set up app store with missions and rockets sub-store using Redux Duck format with each sub-store having its own Redux state slice directory and files.
- Create routes and view components using <NavLink /> for the page navigation links and style active class to indicate which section/page the user is currently on (underline active navigation link). Page links are
   - Rockets page (as the home page)
   - Missions page 
   - Dragons page 
   - Profile Page
- Used Redux createAsyncThunk to fetch data from the SpaceX API upon the first render. The following endpoints are used:
   - Rockets: https://api.spacexdata.com/v3/rockets
   - Missions: https://api.spacexdata.com/v3/missions
   - Dragons: https://api.spacexdata.com/v3/dragons
- Render UI from the store:
  - Use useSelector() Redux Hook to select the state slices and render lists of rockets and missions in corresponding routes
  - Render a list of rockets (based on the design) on the Rockets page.
  - Render a table with the missions' data (based on the design) on the Missions page.
  - Used conditional rendering of components with conditional classes and following KISS, DRY, and YAGNI approach avoid bulk code structure.
- Implement Redux actions and reducers for booking rockets and joining missions
  - Clicking the Reserve Rocket button makes a dispatch to update the store with the corresponding rocket booking action.
  - Clicking the Join Mission button makes a dispatch to update the store with the corresponding mission-joining action.
- Implement Redux actions and reducers for cancelling rockets reservations and leaving missions
  - Clicking the Cancel Reservation button makes a dispatch to update the store with the corresponding rocket booking cancelling action.
  - Clicking the Leave Mission button makes a dispatch to update the store with the corresponding action to leave the mission.
- Render badges to indicate rocket reservation and mission joining
  - Reserved rockets have a Reserved badge upon render
  - Joined missions have an Active Member badge on render while other missions have a NOT A MEMBER` badge on them
- The Profile page shows two-column layouts and lists ONLY the rockets reserved and missions joined by the user.

## Output demo
Screenshot_2

### CODE REVIEW

File: src/redux/missions/missions.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  return { type: START, value: result };
+}
+
+export function LoadMessions(result) {
+  return { type: LOAD, value: result };
+}
+
+export function addremoveMembertoMission(result) {
+  return { type: ADD_REMOVE_MEMBER, value: result };
+}
```
Comment (Chimwemwe-127001) [inline src/redux/missions/missions.js:57]: - [ ] Consider fixing the demonstrated issue below. When a user clicks the "Reserve rocket" button or "Joined Mission" button _[only if your team has 3 members]_, action needs to be dispatched to update the store. You need to get the ID of the reserved rocket and update the state. Remember you mustn't mutate the state. Instead, you need to return a new state object with all rockets, but the selected rocket will have an extra key `reserved` with its value set to `true`.

https://user-images.githubusercontent.com/52143815/193415679-cd260e96-519d-4f00-b91c-8eb1c5c4b554.mp4

File: src/components/pages/Profile.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+            )))}
+          </ListGroup>
+        </div>
+        <div className="profile-rockets">
+          <h2>My Rockets</h2>
+          <ListGroup>
+            {Rockects.map(((rock) => (
+              <ListGroup.Item key={rock.id}>{rock.rocket_name}</ListGroup.Item>
+            )))}
+          </ListGroup>
```
Comment (Chimwemwe-127001) [inline src/components/pages/Profile.js:27]: - [ ] Seeing that you are a trio (team of three), it is imperative that you add a dragon ListGroup here.

**Review by Chimwemwe-127001 — CHANGES_REQUESTED** [review]
# Required Changes ♻️ :warning: 
### Hi @username-forGithub, @lazyCode-404 and @zakiyousufi 

Good job so far! :tada: 

There are some issues that you still need to work on to go to the next project but you are almost there! :pinching_hand: 

so close

## Highlight :+1: 
- No linter errors :heavy_check_mark:
- Repo looks professional :briefcase:
- Solid pull request :scroll:

## Required Changes ♻️
 - Your finished project should work [like this](https://space-travelers-hub-react.netlify.app/)
 
## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

### Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+/* eslint-disable import/no-named-as-default */
+import React from 'react';
+import ReactDOM from 'react-dom/client';
+import { BrowserRouter as Router } from 'react-router-dom';
+import { Provider } from 'react-redux';
+import App from './App';
+import './App.css';
+import store from './Redux/configStore';
```
Comment (phelian23) [inline src/index.js:8]: - [ ] Kindly fix the file reference for importing the store. It should be `./redux/configstore` with lowercase `R`.

File: src/pages/MissionPage.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
+/* eslint-disable jsx-a11y/control-has-associated-label */
+import { React, useEffect } from 'react';
+import { useSelector, useDispatch } from 'react-redux';
+import { fetchMissions } from '../Redux/Missions/missions';
```
Comment (phelian23) [inline src/pages/MissionPage.js:4]: - [ ] kindly fix the file reference here too

File: src/Components/Dragon.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
+import React from 'react';
+import { useDispatch } from 'react-redux';
+import PropTypes from 'prop-types';
+import { CHANGE_STATUE } from '../Redux/dragons/dragons';
```
Comment (phelian23) [inline src/Components/Dragon.js:4]: - [ ] Kindly fix the file reference here too.

File: src/Components/Rocket.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
+import React, { useEffect } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import PropTypes from 'prop-types';
+import { fetchRockets, ChangeStatus } from '../Redux/Rockets/Rockets';
```
Comment (phelian23) [inline src/Components/Rocket.js:4]: - [ ] Kindly fix the file reference here too.

File: src/pages/RocketPage.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import React, { useEffect } from 'react';
+import { useSelector, useDispatch } from 'react-redux';
+import { fetchRockets } from '../Redux/Rockets/Rockets';
```
Comment (phelian23) [inline src/pages/RocketPage.js:3]: - [ ] Kindly fix the file reference here too.

File: src/redux/configStore.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import { configureStore } from '@reduxjs/toolkit';
+import logger from 'redux-logger';
+import thunk from 'redux-thunk';
+import missionsReducer from './Missions/missions';
```
Comment (phelian23) [inline src/redux/configStore.js:4]: - [ ] Kindly fix the file reference here too.

**Review by phelian23 — CHANGES_REQUESTED** [review]
Hi @Team,

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
Requested changes made 👍🏾

 ## Required Changes ♻️

- [ ] Good job working on the requested changes, however, I'm getting an error while starting the app because files are not referenced properly. kindly fix this bug.
Capture


_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/redux/configStore.js (hunk 8 above)
Comment (zakiyousufi) [reply src/redux/configStore.js:4]: Hi @phelian23 
Thanks for your feedback 
Can you please tell me which thing I should change in this file?

Comment (phelian23) [reply src/redux/configStore.js:4]: @zakiyousufi The reference for the missionReducer should be `./missions/missions` with lowercase `m`

File: src/pages/MissionPage.js (hunk 4 above)
Comment (phelian23) [reply src/pages/MissionPage.js:4]: here too it should be `../redux/missions/missions`

File: src/redux/configStore.js (hunk 8 above)
Comment (zakiyousufi) [reply src/redux/configStore.js:4]: @phelian23 
Thanks but after the changing I faced with these errors

spaceship

Comment (zakiyousufi) [reply src/redux/configStore.js:4]: How Can I solve these errors

Comment (phelian23) [reply src/redux/configStore.js:4]: Can you get on a call so we look at it?

Comment (zakiyousufi) [reply src/redux/configStore.js:4]: @phelian23 
Yes, please check your slack or email I will send you an invitation link

**Review by gemmen29 — APPROVED** [review]
Hi Team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 💯 
- No linters errors ✔️ 
- Descriptive pull request ✔️ 
- Descriptive commit messages ✔️ 



## Optional suggestions

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**